File: Task/15-puzzle-game-in-3D/Go/15-puzzle-game-in-3d.go
package main

import (
    "fmt"
    "github.com/gen2brain/raylib-go/raylib"
    "math"
    "math/rand"
    "strconv"
    "time"
)

var palette = []rl.Color{
    rl.Blue,
    rl.Green,
    rl.Red,
    rl.SkyBlue,
    rl.Magenta,
    rl.Gray,
    rl.Lime,
    rl.Purple,
    rl.Violet,
    rl.Pink,
    rl.Gold,
    rl.Orange,
    rl.Maroon,
    rl.Beige,
    rl.Brown,
    rl.RayWhite,
}

var (
    screenWidth  = int32(960)
    screenHeight = int32(840)
    radius       = screenHeight / 14
    fontSize     = 2 * radius / 5
    angle        = math.Pi / 6
    incr         = 2 * angle
    blank        = 15
    moves        = 0
    gameOver     = false
)

var (
    centers [16]rl.Vector2
    cubes   [16]int
)

func init() {
    for i := 0; i < 16; i++ {
        cubes[i] = i
    }
}

func drawCube(n, pos int) {
    r := float32(radius)
    rl.DrawPoly(centers[pos], 6, r, 0, palette[n])
    cx, cy := centers[pos].X, centers[pos].Y
    for i := 1; i <= 5; i += 2 {
        fi := float64(i)
        vx := int32(r*float32(math.Cos(angle+fi*incr)) + cx)
        vy := int32(r*float32(math.Sin(angle+fi*incr)) + cy)
        rl.DrawLine(int32(cx), int32(cy), vx, vy, rl.Black)
    }
    ns := ""
    if n < 15 || gameOver {
        ns = strconv.Itoa(n + 1)
    }
    hr, er, tqr := r/2, r/8, 0.75*r
    rl.DrawText(ns, int32(cx+hr-er), int32(cy), fontSize, rl.RayWhite)
    rl.DrawText(ns, int32(cx-hr-er), int32(cy), fontSize, rl.RayWhite)
    rl.DrawText(ns, int32(cx-er), int32(cy-tqr), fontSize, rl.RayWhite)
}

func updateGame() {
    if gameOver {
        return
    } else if rl.IsKeyPressed(rl.KeyLeft) {
        if blank%4 != 0 {
            cubes[blank], cubes[blank-1] = cubes[blank-1], cubes[blank]
            blank--
            moves++
        }
    } else if rl.IsKeyPressed(rl.KeyRight) {
        if (blank+1)%4 != 0 {
            cubes[blank], cubes[blank+1] = cubes[blank+1], cubes[blank]
            blank++
            moves++
        }
    } else if rl.IsKeyPressed(rl.KeyUp) {
        if blank > 3 {
            cubes[blank], cubes[blank-4] = cubes[blank-4], cubes[blank]
            blank -= 4
            moves++
        }
    } else if rl.IsKeyPressed(rl.KeyDown) {
        if blank < 12 {
            cubes[blank], cubes[blank+4] = cubes[blank+4], cubes[blank]
            blank += 4
            moves++
        }
    }
}

func completed() bool {
    for i := 0; i < 16; i++ {
        if cubes[i] != i {
            return false
        }
    }
    palette[15] = rl.DarkGreen
    gameOver = true
    return true
}

func main() {
    rand.Seed(time.Now().UnixNano())
    rand.Shuffle(15, func(i, j int) {
        cubes[i], cubes[j] = cubes[j], cubes[i]
    })
    rl.InitWindow(screenWidth, screenHeight, "15-puzzle game using 3D cubes")
    rl.SetTargetFPS(60)
    var x, y = float32(screenWidth) / 10, float32(radius)
    for i := 0; i < 4; i++ {
        cx := 2 * x * float32(i+1)
        for j := 0; j < 4; j++ {
            cy := (x + y) * float32(j+1)
            centers[j*4+i] = rl.NewVector2(cx, cy)
        }
    }

    for !rl.WindowShouldClose() {
        rl.BeginDrawing()
        rl.ClearBackground(rl.Black)
        for i := 0; i < 16; i++ {
            drawCube(cubes[i], i)
        }

        if !completed() {
            m := fmt.Sprintf("Moves = %d", moves)
            rl.DrawText(m, 4*int32(x), 13*int32(y), fontSize, rl.RayWhite)
        } else {
            m := fmt.Sprintf("You've completed the puzzle in %d moves!", moves)
            rl.DrawText(m, 3*int32(x), 13*int32(y), fontSize, rl.RayWhite)
        }
        rl.EndDrawing()
        updateGame()
    }

    rl.CloseWindow()
}

File: Task/15-puzzle-game-in-3D/Julia/15-puzzle-game-in-3d.jl
using Random
using Raylib
using StaticArrays

# Define color palette
const palette = [
    Raylib.BLUE,
    Raylib.GREEN,
    Raylib.RED,
    Raylib.SKYBLUE,
    Raylib.MAGENTA,
    Raylib.GRAY,
    Raylib.LIME,
    Raylib.PURPLE,
    Raylib.VIOLET,
    Raylib.PINK,
    Raylib.GOLD,
    Raylib.ORANGE,
    Raylib.MAROON,
    Raylib.BEIGE,
    Raylib.BROWN,
    Raylib.RAYWHITE
]

# Constants
const screenWidth = Int32(960)
const screenHeight = Int32(840)
const radius = screenHeight ÷ 14
const fontSize = 2 * radius ÷ 5
const angle = π / 6
const incr = 2 * angle
const blank = Ref{Int}(15)  # Using Ref for mutable global
const moves = Ref{Int}(0)
const gameover = Ref{Bool}(false)

# Arrays
const centers = Vector{SVector{2}}(undef, 16)
const cubes = Vector{Int}(0:15)  # Initialized with 0 to 15


function drawcube(n::Int, pos::Int)
    """ Draw a single cube numbered n at position pos """
    r = Float32(radius)
    Raylib.DrawPoly(centers[pos+1], 6, r, 0, palette[n+1])
    cx, cy = centers[pos+1].x, centers[pos+1].y
    for i in 1:2:5
        fi = Float64(i)
        vx = Int32(round(r * cos(angle + fi * incr) + cx))
        vy = Int32(round(r * sin(angle + fi * incr) + cy))
        Raylib.DrawLine(Int32(round(cx)), Int32(round(cy)), vx, vy, Raylib.BLACK)
    end
    ns = (n < 15 || gameover[]) ? string(n + 1) : ""
    hr, er, tqr = r / 2, r / 8, 0.75 * r
    Raylib.DrawText(ns, Int32(round(cx + hr - er)), Int32(round(cy)), fontSize, Raylib.RAYWHITE)
    Raylib.DrawText(ns, Int32(round(cx - hr - er)), Int32(round(cy)), fontSize, Raylib.RAYWHITE)
    Raylib.DrawText(ns, Int32(round(cx - er)), Int32(round(cy - tqr)), fontSize, Raylib.RAYWHITE)
end

function update()
    """ Update game state based on keyboard input """
    gameover[] && return  # If game is over, stop taking moves
    if Raylib.IsKeyPressed(Int32(Raylib.KEY_LEFT))
        if blank[] % 4 != 0
            cubes[blank[]+1], cubes[blank[]] = cubes[blank[]], cubes[blank[]+1]
            blank[] -= 1
            moves[] += 1
        end
    elseif Raylib.IsKeyPressed(Int32(Raylib.KEY_RIGHT))
        if (blank[] + 1) % 4 != 0
            cubes[blank[]+1], cubes[blank[]+2] = cubes[blank[]+2], cubes[blank[]+1]
            blank[] += 1
            moves[] += 1
        end
    elseif Raylib.IsKeyPressed(Int32(Raylib.KEY_UP))
        if blank[] > 3
            cubes[blank[]+1], cubes[blank[]-3] = cubes[blank[]-3], cubes[blank[]+1]
            blank[] -= 4
            moves[] += 1
        end
    elseif Raylib.IsKeyPressed(Int32(Raylib.KEY_DOWN))
        if blank[] < 12
            cubes[blank[]+1], cubes[blank[]+5] = cubes[blank[]+5], cubes[blank[]+1]
            blank[] += 4
            moves[] += 1
        end
    end
end


function completed()
    """ returns true if puzzle is completed """
    for i in 1:16
        cubes[i] != i - 1 && return false
    end
    palette[16] = Raylib.DARKGREEN
    gameover[] = true
    return true
end

function puzzle3dcubes()
    """ Main function to run the '15-puzzle' game using 3D drawn cubes """
    # Initialize cubes
    for i in 1:16
        cubes[i] = i - 1
    end
    shuffle!(view(cubes, 1:15))  # Shuffle first 15 elements
    Raylib.InitWindow(screenWidth, screenHeight, "15-puzzle game using 3D cubes")
    Raylib.SetTargetFPS(60)

    x, y = Float32(screenWidth) / 10, Float32(radius)
    for i in 0:3
        cx = 2 * x * (i + 1)
        for j in 0:3
            cy = (x + y) * (j + 1)
            centers[j*4 + i + 1] = SVector{2}([cx, cy])
        end
    end

    while !Raylib.WindowShouldClose()
        Raylib.BeginDrawing()
        Raylib.ClearBackground(Raylib.BLACK)
        for i in 0:15
            drawcube(cubes[i+1], i)
        end

        if !completed()
            m = "Moves = $(moves[])"
            Raylib.DrawText(m, Int32(4 * x), Int32(13 * y), fontSize, Raylib.RAYWHITE)
        else
            m = "You've completed the puzzle in $(moves[]) moves!"
            Raylib.DrawText(m, Int32(3 * x), Int32(13 * y), fontSize, Raylib.RAYWHITE)
        end
        Raylib.EndDrawing()
        update()
    end

    Raylib.CloseWindow()
end

# Run the game
puzzle3dcubes()
